### ISSUE TITLE
500 unknown error when binding route to app in org with quota of small app instance mem

### ISSUE THREAD

**Issue by dkoper** [cloudfoundry/cloud_controller_ng#864]
## Issue

App push failed with an unknown error from CC:
```
D:\Users\koperd\Downloads\bin>cf p app  -p t\n --no-start
Creating app app in org zeromem / space dev as admin...
OK

Creating route app.example.com...
OK

Binding app.example.com to app...
FAILED
Server error, status code: 500, error code: 10001, message: An unknown error occurred.
```

## Context

In an org with 0M quota for "Maximum amount of memory an application instance can have".
Does not seem specific to particular version: reproduced using cf CLI 6.29.0 and CF v254 as well as latest CC 2.90.0.
Trace:
```
REQUEST: [2017-08-02T15:02:54+10:00]
PUT /v2/apps/df47aaf3-8a7c-4312-9701-2fc1d6c7e212/routes/c412b319-f262-4f95-9a84-7ad8a74776f2 HTTP/1.1
Host: api.example.com
Accept: application/json
Authorization: [PRIVATE DATA HIDDEN]
Connection: close
Content-Type: application/json
User-Agent: go-cli 6.29.0+f8f0b73.2017-07-31 / windows



RESPONSE: [2017-08-02T15:02:56+10:00]
HTTP/1.1 500 Internal Server Error
Connection: close
Content-Length: 99
Alt-Svc: clear
Content-Type: application/json;charset=utf-8
Date: Wed, 02 Aug 2017 05:02:55 GMT
Server: nginx
Via: 1.1 google
X-Content-Type-Options: nosniff
X-Vcap-Request-Id: 0ee3912d-8bde-41bb-6c4a-abf3c2772341
X-Vcap-Request-Id: 5004de9e-f845-4059-4fee-cef1396bf4fe::fa6c2491-1c19-44db-ba3d-6329184ce6d1

{
  "error_code": "UnknownError",
  "description": "An unknown error occurred.",
  "code": 10001
}
```
## Steps to Reproduce

Inline for easy copy & pasting:
```
cf create-quota zeromem -i 0M -r 5
cf create-org zeromem -q zeromem
cf t -o zeromem
cf create-space dev
cf t -s dev
cf p app --no-start -m 256M
```

## Expected result
Operation to succeed (as I'm not starting the app), or possibly memory error from staging.

## Current result
Push failing with unknown error.

**Comment by tcdowney** [comment]
Here's the actual error that's being raised under the hood:

```
{
   "timestamp":1502216411.4108799,
   "message":"   Request failed:500:{
      \"error_code\"=>\"UnknownError\",
      \"description\"=>\"An unknown error occurred.\",
      \"code\"=>10001,
      \"test_mode_info\"=>      {
         \"description\"=>\"memory instance_memory_limit_exceeded\",
         \"error_code\"=>\"CF-InvalidRouteMapping\",
         \"backtrace\"=>         [
            \"/var/vcap/packages/cloud_controller_ng/cloud_controller_ng/app/actions/v2/route_mapping_create.rb:62:in `rescue in add'\",
              ...
         ]
      }
   }
}
```

We end up updating the `process` that the route is being mapped to [here](https://github.com/cloudfoundry/cloud_controller_ng/blob/0f6d81989aa35d74cfec6a65f01172655abe2da2/lib/cloud_controller/process_route_handler.rb#L8) and that causes it to run through the various validations on the process.  Here it looks like it's hitting the [AppMaxInstanceMemoryPolicy](https://github.com/cloudfoundry/cloud_controller_ng/blob/0f6d81989aa35d74cfec6a65f01172655abe2da2/app/models/runtime/process_model.rb#L215) validation and that is ultimately being converted into a generic `InvalidRouteMapping` error which is not being handled by the "add route" endpoint.

@dkoper Using your reproduction steps, if you run `cf p app --no-start -m 256M -v` again after doing all of this once, it does surface a more helpful error since it fails in the actual app update step and the error is handled.

```
PUT /v2/apps/7b436245-87b8-4fc9-a380-73888bc5f117?async=true&inline-relations-depth=1 HTTP/1.1

{
  "description": "You have exceeded the instance memory limit for your organization's quota.",
  "error_code": "CF-QuotaInstanceMemoryLimitExceeded",
  "code": 100007
}

FAILED
Server error, status code: 400, error code: 100007, message: You have exceeded the instance memory limit for your organization's quota.
```

@zrob is the behavior of the app update endpoint correct?  The app create endpoint allows the request, so it seems like we either update `#create` to reject it as well or update `#update` to allow it since the app is stopped.

**Comment by matt-royal** [comment]
@zrob We did some more research and dug into the cause of this problem. The details are in the tracker story.

**Comment by andrew-edgar** [comment]
+1 for this defect as we also saw it when pushing and exceeding the quota (for example,

cf push app -m 100G 

If the app does not yet exist we fail with a 500.   If we are updating an existing app to -m 100G the failure is the correct 400 exceeded memory limit.

Would appreciate an update of this defect and when it will be addressed.

Thanks,

Andrew Edgar

**Comment by Gerg** [comment]
Thanks for raising this issue. We pushed a fix for this issue here: https://github.com/cloudfoundry/cloud_controller_ng/commit/118fc7b3591e480744591f07ae21e0976317fff8

This should be included in `capi-release` 1.44.0. Please re-open this issue if you continue to see this problem going forward.

Thanks,
Greg
